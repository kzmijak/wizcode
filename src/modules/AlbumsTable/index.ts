export { AlbumsTable } from "./src/AlbumsTable";
