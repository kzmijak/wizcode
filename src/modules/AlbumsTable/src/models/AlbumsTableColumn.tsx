export const albumsTableColumnConsts = [
  "Index",
  "Image",
  "Title",
  "Artist",
] as const;

export type AlbumsTableColumn = (typeof albumsTableColumnConsts)[number];
