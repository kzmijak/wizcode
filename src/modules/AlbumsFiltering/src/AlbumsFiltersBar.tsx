import { FC } from "react";
import { Search } from "./components/Search";
import { CategoryChips } from "modules/CategoryFilter";
import { Category } from "$enums/Category";

type AlbumsFiltersBarProps = {
  search: string;
  onSearchChange: (value: string) => void;
  allCategories: Category[];
};

export const AlbumsFiltersBar: FC<AlbumsFiltersBarProps> = ({
  onSearchChange,
  search,
  allCategories,
}) => (
  <section>
    <Search value={search} onChange={onSearchChange} />
    <CategoryChips allCategories={allCategories} />
  </section>
);
