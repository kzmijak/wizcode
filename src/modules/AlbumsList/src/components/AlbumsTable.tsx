import { Table } from "@mui/joy";
import { AlbumModel } from "models/Album";
import { FC } from "react";

type AlbumsTableProps = {
  rows: AlbumModel[];
};

export const AlbumsTable: FC<AlbumsTableProps> = ({ rows }) => {
  return (
    <Table
      aria-label="albums table"
      borderAxis="none"
      color="neutral"
      size="md"
      stickyHeader
      stripe="odd"
      variant="plain"
    >
      <thead>
        <tr>
          <th>#</th>
          <th>Image</th>
          <th>Title</th>
          <th>Artist</th>
        </tr>
      </thead>
      <tbody>
        {rows.map((row, index) => (
          <tr key={row.id}>
            <td>{index + 1}</td>
            <td>
              <img src={row.imageUrl} alt={`${row.title} album image`} />
            </td>
            <td>{row.title}</td>
            <td>{row.artistName}</td>
          </tr>
        ))}
      </tbody>
    </Table>
  );
};
