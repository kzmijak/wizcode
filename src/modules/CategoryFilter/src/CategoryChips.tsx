import { Sheet, Stack } from "@mui/joy";
import { FC } from "react";
import { CategoryChipItem } from "./components/CategoryChipItem";
import { Category } from "$enums/Category";

export type CategoryChipsProps = {
  allCategories: Category[];
};

export const CategoryChips: FC<CategoryChipsProps> = ({ allCategories }) => {
  return (
    <Sheet>
      <Stack
        direction="row"
        spacing={0.5}
        maxWidth="100%"
        flexWrap="wrap"
        padding={1}
      >
        {allCategories.map((category) => (
          <CategoryChipItem key={category} category={category} />
        ))}
      </Stack>
    </Sheet>
  );
};
