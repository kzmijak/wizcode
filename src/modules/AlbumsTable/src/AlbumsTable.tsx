import { Table } from "@mui/joy";
import { AlbumModel } from "models/Album";
import { FC, useState } from "react";
import { Header } from "./components/Header";
import { filterRowsByText } from "./utils/filterRowsByText";
import { albumsTableColumnConsts } from "./models/AlbumsTableColumn";
import { ColumnHeader } from "./components/cells/ColumnHeader";

type AlbumsTableProps = {
  rows: AlbumModel[];
};

export const AlbumsTable: FC<AlbumsTableProps> = ({ rows }) => {
  const [search, setSearch] = useState("");

  const displayRows = filterRowsByText(rows, search);

  return (
    <>
      <Header search={search} onSearchChange={setSearch} />
      <Table
        aria-label="albums table"
        borderAxis="none"
        color="neutral"
        size="md"
        stickyHeader
        stripe="odd"
        variant="plain"
      >
        <thead>
          <tr>
            {albumsTableColumnConsts.map((column) => (
              <ColumnHeader column={column} />
            ))}
          </tr>
        </thead>
        <tbody>
          {displayRows.map((row) => (
            <tr key={row.id}>
              <td>{rows.indexOf(row) + 1}</td>
              <td>
                <img src={row.imageUrl} alt={`${row.title} album image`} />
              </td>
              <td>{row.title}</td>
              <td>{row.artistName}</td>
            </tr>
          ))}
        </tbody>
      </Table>
    </>
  );
};
