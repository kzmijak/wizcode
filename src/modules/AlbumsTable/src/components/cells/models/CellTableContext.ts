import { AlbumsTableRow } from "modules/AlbumsTable/src/models/AlbumsTableRow";

export type CellTableContext<TValueType = string> = {
  value: TValueType;
  currentRow: AlbumsTableRow;
  rows: AlbumsTableRow[];
};
