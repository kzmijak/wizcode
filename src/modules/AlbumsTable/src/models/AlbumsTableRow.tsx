import { AlbumModel } from "models/Album";

export type AlbumsTableRow = AlbumModel;
