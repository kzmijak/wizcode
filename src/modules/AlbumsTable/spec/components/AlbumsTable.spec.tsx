import { describe, it, expect } from "vitest";
import { render, screen } from "@testing-library/react";
import { AlbumsTable } from "../../src/AlbumsTable";
import { AlbumModel } from "models/Album";

const imageUrl = "https://picsum.photos";

const rows: AlbumModel[] = [
  {
    id: "1",
    artistName: "Artist 1",
    title: "Title 1",
    imageUrl,
  },
  {
    id: "2",
    artistName: "Artist 2",
    title: "Title 2",
    imageUrl,
  },
  {
    id: "3",
    artistName: "Artist 3",
    title: "Title 3",
    imageUrl,
  },
];

describe("AlbumsTable", () => {
  const Component = () => <AlbumsTable rows={rows} />;

  it("should match snapshot", () => {
    expect(render(<Component />)).toMatchSnapshot();
  });

  it("should render table with headers", () => {
    render(<Component />);

    expect(screen.getByText("#")).toBeVisible();
    expect(screen.getByText("Image")).toBeVisible();
    expect(screen.getByText("Title")).toBeVisible();
    expect(screen.getByText("Artist")).toBeVisible();
  });

  it("should render adequate labels", () => {
    render(<Component />);

    expect(screen.getByRole("cell", { name: "Artist 1" })).toBeVisible();
    expect(screen.getByRole("cell", { name: "Title 1" })).toBeVisible();
  });

  it("should render properly described image based on the album title", () => {
    render(<Component />);

    const image = screen.queryByAltText("Title 1 album image");
    expect(image).toBeInTheDocument();
    expect(image).toHaveAttribute("src", imageUrl);
  });

  it("should render as many rows as there were passed in prop", () => {
    render(<Component />);

    const lastIndexCell = screen.queryByRole("cell", {
      name: rows.length.toString(),
    });

    expect(lastIndexCell).toBeInTheDocument();
  });

  it("should not render more rows than passed in prop", () => {
    render(<Component />);

    const overflowingIndexCell = screen.queryByRole("cell", {
      name: (rows.length + 1).toString(),
    });

    expect(overflowingIndexCell).not.toBeInTheDocument();
  });
});
