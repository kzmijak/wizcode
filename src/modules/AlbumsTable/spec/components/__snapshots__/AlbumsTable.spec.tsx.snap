// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AlbumsTable > should match snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <section>
        <div
          class="MuiFormControl-root MuiFormControl-vertical MuiFormControl-sizeMd css-akqkp6-JoyFormControl-root"
        >
          <label
            class="MuiFormLabel-root css-7g4fyo-JoyFormLabel-root"
            for=":r0:"
            id=":r0:-label"
          >
            Search
          </label>
          <div
            class="MuiInput-root MuiInput-variantOutlined MuiInput-colorNeutral MuiInput-sizeMd MuiInput-formControl css-1vip6bn-JoyInput-root"
          >
            <input
              class="MuiInput-input css-1gw9vc6-JoyInput-input"
              id=":r0:"
              placeholder="Type here..."
              type="search"
              value=""
            />
          </div>
        </div>
      </section>
      <table
        aria-label="albums table"
        class="MuiTable-root MuiTable-stickyHeader MuiTable-borderAxisNone MuiTable-variantPlain MuiTable-colorNeutral MuiTable-sizeMd css-rokh38-JoyTable-root"
      >
        <thead>
          <tr>
            <th
              class="MuiBox-root css-1x769y4"
            >
              <p
                class="MuiTypography-root MuiTypography-inherit css-1czpeps-JoyTypography-root"
              >
                #
              </p>
            </th>
            <th
              class="MuiBox-root css-1bvg7gj"
            >
              <p
                class="MuiTypography-root MuiTypography-inherit css-1czpeps-JoyTypography-root"
              >
                Image
              </p>
            </th>
            <th
              class="MuiBox-root css-crm8vv"
            >
              <p
                class="MuiTypography-root MuiTypography-inherit css-1czpeps-JoyTypography-root"
              >
                Title
              </p>
            </th>
            <th
              class="MuiBox-root css-j3puv3"
            >
              <p
                class="MuiTypography-root MuiTypography-inherit css-1czpeps-JoyTypography-root"
              >
                Artist
              </p>
            </th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td>
              1
            </td>
            <td>
              <img
                alt="Title 1 album image"
                src="https://picsum.photos"
              />
            </td>
            <td>
              Title 1
            </td>
            <td>
              Artist 1
            </td>
          </tr>
          <tr>
            <td>
              2
            </td>
            <td>
              <img
                alt="Title 2 album image"
                src="https://picsum.photos"
              />
            </td>
            <td>
              Title 2
            </td>
            <td>
              Artist 2
            </td>
          </tr>
          <tr>
            <td>
              3
            </td>
            <td>
              <img
                alt="Title 3 album image"
                src="https://picsum.photos"
              />
            </td>
            <td>
              Title 3
            </td>
            <td>
              Artist 3
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </body>,
  "container": <div>
    <section>
      <div
        class="MuiFormControl-root MuiFormControl-vertical MuiFormControl-sizeMd css-akqkp6-JoyFormControl-root"
      >
        <label
          class="MuiFormLabel-root css-7g4fyo-JoyFormLabel-root"
          for=":r0:"
          id=":r0:-label"
        >
          Search
        </label>
        <div
          class="MuiInput-root MuiInput-variantOutlined MuiInput-colorNeutral MuiInput-sizeMd MuiInput-formControl css-1vip6bn-JoyInput-root"
        >
          <input
            class="MuiInput-input css-1gw9vc6-JoyInput-input"
            id=":r0:"
            placeholder="Type here..."
            type="search"
            value=""
          />
        </div>
      </div>
    </section>
    <table
      aria-label="albums table"
      class="MuiTable-root MuiTable-stickyHeader MuiTable-borderAxisNone MuiTable-variantPlain MuiTable-colorNeutral MuiTable-sizeMd css-rokh38-JoyTable-root"
    >
      <thead>
        <tr>
          <th
            class="MuiBox-root css-1x769y4"
          >
            <p
              class="MuiTypography-root MuiTypography-inherit css-1czpeps-JoyTypography-root"
            >
              #
            </p>
          </th>
          <th
            class="MuiBox-root css-1bvg7gj"
          >
            <p
              class="MuiTypography-root MuiTypography-inherit css-1czpeps-JoyTypography-root"
            >
              Image
            </p>
          </th>
          <th
            class="MuiBox-root css-crm8vv"
          >
            <p
              class="MuiTypography-root MuiTypography-inherit css-1czpeps-JoyTypography-root"
            >
              Title
            </p>
          </th>
          <th
            class="MuiBox-root css-j3puv3"
          >
            <p
              class="MuiTypography-root MuiTypography-inherit css-1czpeps-JoyTypography-root"
            >
              Artist
            </p>
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            1
          </td>
          <td>
            <img
              alt="Title 1 album image"
              src="https://picsum.photos"
            />
          </td>
          <td>
            Title 1
          </td>
          <td>
            Artist 1
          </td>
        </tr>
        <tr>
          <td>
            2
          </td>
          <td>
            <img
              alt="Title 2 album image"
              src="https://picsum.photos"
            />
          </td>
          <td>
            Title 2
          </td>
          <td>
            Artist 2
          </td>
        </tr>
        <tr>
          <td>
            3
          </td>
          <td>
            <img
              alt="Title 3 album image"
              src="https://picsum.photos"
            />
          </td>
          <td>
            Title 3
          </td>
          <td>
            Artist 3
          </td>
        </tr>
      </tbody>
    </table>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
